src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py

```diff
@@ -41,10 +41,41 @@ def _generate_bias_rules(self, bias_data: BiasData) -> List[BaseRule]:
                 },
                 "active": True,
                 "id": bias_data["id"],
-            }
+            }  # type:ignore
+        ]
+
+
+class BoostEnvironmentsRulesGeneratorV2(BiasRulesGenerator):
+    def _generate_bias_rules(self, bias_data: BiasData) -> List[BaseRule]:
+        return [
+            {
+                "samplingValue": {
+                    "type": "sampleRate",
+                    "value": 1.0,
+                },
+                "type": "trace",
+                "condition": {
+                    "op": "or",
+                    "inner": [
+                        {
+                            "op": "glob",
+                            "name": "trace.environment",
+                            "value": ENVIRONMENT_GLOBS,
+                            "options": {"ignoreCase": True},
+                        }
+                    ],
+                },
+                "active": True,
+                "id": bias_data["id"],
+            }  # type:ignore
```

Comment:
Please add proper type hints for the new config format. Though they are not a guarantee that Relay can parse them, they help us catch obvious typos by comparing the rust types to the python types side-by-side.

Reply:
I did add them, you can see in the PR the `SamplingValue` in utils. The reason for `# type:ignore` is that mypy doesn't seem to like `Optional` fields. I have tried with `NotRequired` and it doesn't work.

Do you have any suggestions?

Reply:
I would change the signature of this function (and all other V2 generators) to `List[RuleV2]`, and define `RuleV2` as

```python
class RuleV1(TypedDict):  # formerly known as BaseRule
   ...
   sampleRate: float

class RuleV2(TypedDict):
    ...
    samplingValue: SamplingValue
```

Reply:
Yep, will do! Looking at python fake typing it looks like there is no issue in changing types during method overriding right?

Reply:
You could make `BiasRulesGenerator` generic over the rule version, see for example https://github.com/getsentry/sentry/blob/a9f0a281ef2534d86c54b70cc85238f60700feec/src/sentry/utils/kvstore/abstract.py#L5-L9

Reply:
Damn, Python type hints is its own world of trying to be something it isn't going to be.